### PR TITLE
Rule cognitive complexity

### LINKED ISSUE #268: New rule: cognitive complexity
It could be nice to have a rule to check for Cognitive Complexity (see [Cognitive complexity - A new way of measuring understandability](https://www.sonarsource.com/docs/CognitiveComplexity.pdf)), a complexity measure complementary to cyclomatic complexity.

### PR DESCRIPTION
**Pull request by chavacava**
Closes #268

### CODE REVIEW

File: RULES_DESCRIPTIONS.md [hunk 1]
```diff
@@ -122,6 +123,19 @@ The garbage collector can be configured through environment variables as describ
 
 _Configuration_: N/A
 
+## cognitive-complexity
+
+_Description_: [Cognitive complexity](https://www.sonarsource.com/resources/white-papers/cognitive-complexity.html) is a measure of how hard code is to understand. Enforcing a maximum complexity per function helps to keep code readable and maintainable.
```
Comment (mgechev) [inline RULES_DESCRIPTIONS.md]: Maybe one sentence on how it differs from cyclomatic complexity?

File: fixtures/cognitive-complexity.go [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+	if ok := f(); ok { // +1
+		return "it's okay"
+	} else {
+		return "it's NOT okay!"
+	}
+}
+
+// Test Boolean expr
+func h(a, b, c, d, e, f bool) bool { // MATCH /function h has cognitive complexity 2 (> max enabled 0)/
+	return a && b && c || d || e && f //FIXME: complexity should be 3?
```
Comment (mgechev) [inline fixtures/cognitive-complexity.go]: Yes, I think it should be 3 as well:

>Cognitive complexity increments for each new sequence of like operators

Comment (chavacava) [reply fixtures/cognitive-complexity.go]: Doubt: 
`a && !(b || c)` 
`a && !(b && c)`

have both he same complexity (2)?

File: RULES_DESCRIPTIONS.md (hunk 1 above)
Comment (chavacava) [reply RULES_DESCRIPTIONS.md]: Done.

File: fixtures/cognitive-complexity.go (hunk 2 above)
Comment (chavacava) [reply fixtures/cognitive-complexity.go]: I re-wrote the computation of complexity of binary expressions. Now it is in line with the definition

Comment (mgechev) [reply fixtures/cognitive-complexity.go]: >Doubt: 
>`a && !(b || c)` 
>`a && !(b && c)`
>
>have both he same complexity (2)?

I think so.

File: rule/cognitive-complexity.go [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+	calculator := binExprComplexityCalculator{currentOp: []token.Token{}}
+
+	astutil.Apply(n, calculator.pre(), calculator.post())
+
+	return calculator.complexity
+}
+
+type binExprComplexityCalculator struct {
+	complexity    int
+	currentOp     []token.Token // stack of bool operators
```
Comment (mgechev) [inline rule/cognitive-complexity.go]: It's a list of operations, right? Can we call it `ops`?

File: rule/cognitive-complexity.go [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+	subexpStarted bool
+}
+
+func (becc *binExprComplexityCalculator) pre() astutil.ApplyFunc {
+	return func(c *astutil.Cursor) bool {
```
Comment (mgechev) [inline rule/cognitive-complexity.go]: Do we need the higher order function here? Can we have:

```go
func (becc *binExprComplexityCalculator) pre(c *astutil.Cursor) bool {
  // ...
}
```

Comment (chavacava) [reply rule/cognitive-complexity.go]: Not really, you are right. Done

Comment (chavacava) [reply rule/cognitive-complexity.go]: Not really, you are right. Done

**Pull request merged**